Some 32-bit platforms now need to link to libatomic: Undefined symbols: "___atomic_store_8"

```
[100%] Linking CXX executable tester
cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_math_blaspp/blaspp/work/build/test && /opt/local/bin/cmake -E cmake_link_script CMakeFiles/tester.dir/link.txt --verbose=ON
/opt/local/bin/g++-mp-14 -pipe -Os -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -arch ppc -mmacosx-version-min=10.6 -Wl,-search_paths_first -Wl,-headerpad_max_install_names -L/opt/local/lib -Wl,-headerpad_max_install_names CMakeFiles/tester.dir/test.cc.o CMakeFiles/tester.dir/test_util.cc.o CMakeFiles/tester.dir/test_asum.cc.o CMakeFiles/tester.dir/test_axpy.cc.o CMakeFiles/tester.dir/test_batch_gemm.cc.o CMakeFiles/tester.dir/test_batch_hemm.cc.o CMakeFiles/tester.dir/test_batch_her2k.cc.o CMakeFiles/tester.dir/test_batch_herk.cc.o CMakeFiles/tester.dir/test_batch_symm.cc.o CMakeFiles/tester.dir/test_batch_syr2k.cc.o CMakeFiles/tester.dir/test_batch_syrk.cc.o CMakeFiles/tester.dir/test_batch_trmm.cc.o CMakeFiles/tester.dir/test_batch_trsm.cc.o CMakeFiles/tester.dir/test_copy.cc.o CMakeFiles/tester.dir/test_dot.cc.o CMakeFiles/tester.dir/test_dotu.cc.o CMakeFiles/tester.dir/test_error.cc.o CMakeFiles/tester.dir/test_gemm.cc.o CMakeFiles/tester.dir/test_gemv.cc.o CMakeFiles/tester.dir/test_ger.cc.o CMakeFiles/tester.dir/test_geru.cc.o CMakeFiles/tester.dir/test_hemm.cc.o CMakeFiles/tester.dir/test_hemv.cc.o CMakeFiles/tester.dir/test_her.cc.o CMakeFiles/tester.dir/test_her2.cc.o CMakeFiles/tester.dir/test_her2k.cc.o CMakeFiles/tester.dir/test_herk.cc.o CMakeFiles/tester.dir/test_iamax.cc.o CMakeFiles/tester.dir/test_max.cc.o CMakeFiles/tester.dir/test_memcpy.cc.o CMakeFiles/tester.dir/test_memcpy_2d.cc.o CMakeFiles/tester.dir/test_nrm2.cc.o CMakeFiles/tester.dir/test_rot.cc.o CMakeFiles/tester.dir/test_rotg.cc.o CMakeFiles/tester.dir/test_rotm.cc.o CMakeFiles/tester.dir/test_rotmg.cc.o CMakeFiles/tester.dir/test_scal.cc.o CMakeFiles/tester.dir/test_swap.cc.o CMakeFiles/tester.dir/test_symm.cc.o CMakeFiles/tester.dir/test_symv.cc.o CMakeFiles/tester.dir/test_syr.cc.o CMakeFiles/tester.dir/test_syr2.cc.o CMakeFiles/tester.dir/test_syr2k.cc.o CMakeFiles/tester.dir/test_syrk.cc.o CMakeFiles/tester.dir/test_trmm.cc.o CMakeFiles/tester.dir/test_trmv.cc.o CMakeFiles/tester.dir/test_trsm.cc.o CMakeFiles/tester.dir/test_trsv.cc.o CMakeFiles/tester.dir/cblas_wrappers.cc.o CMakeFiles/tester.dir/lapack_wrappers.cc.o CMakeFiles/tester.dir/test_batch_gemm_device.cc.o CMakeFiles/tester.dir/test_batch_hemm_device.cc.o CMakeFiles/tester.dir/test_batch_her2k_device.cc.o CMakeFiles/tester.dir/test_batch_herk_device.cc.o CMakeFiles/tester.dir/test_schur_gemm.cc.o CMakeFiles/tester.dir/test_batch_symm_device.cc.o CMakeFiles/tester.dir/test_batch_syr2k_device.cc.o CMakeFiles/tester.dir/test_batch_syrk_device.cc.o CMakeFiles/tester.dir/test_batch_trmm_device.cc.o CMakeFiles/tester.dir/test_batch_trsm_device.cc.o CMakeFiles/tester.dir/test_axpy_device.cc.o CMakeFiles/tester.dir/test_dot_device.cc.o CMakeFiles/tester.dir/test_dotu_device.cc.o CMakeFiles/tester.dir/test_nrm2_device.cc.o CMakeFiles/tester.dir/test_scal_device.cc.o CMakeFiles/tester.dir/test_swap_device.cc.o CMakeFiles/tester.dir/test_copy_device.cc.o CMakeFiles/tester.dir/test_gemm_device.cc.o CMakeFiles/tester.dir/test_hemm_device.cc.o CMakeFiles/tester.dir/test_her2k_device.cc.o CMakeFiles/tester.dir/test_herk_device.cc.o CMakeFiles/tester.dir/test_symm_device.cc.o CMakeFiles/tester.dir/test_syr2k_device.cc.o CMakeFiles/tester.dir/test_syrk_device.cc.o CMakeFiles/tester.dir/test_trmm_device.cc.o CMakeFiles/tester.dir/test_trsm_device.cc.o -o tester  -Wl,-rpath,/opt/local/lib ../_deps/testsweeper-build/libtestsweeper.1.0.0.dylib ../libblaspp.1.0.0.dylib /opt/local/lib/gcc14/libgomp.dylib -framework Accelerate
Undefined symbols:
  "___atomic_store_8", referenced from:
      __Z10setup_PAPIPi in test.cc.o
ld: symbol(s) not found
collect2: error: ld returned 1 exit status
make[2]: *** [test/tester] Error 1
```

This is not OS-specific issue, it will fail likewise on *BSD and Linux on certain 32-bit archs (arm, ppc, likely mips).

This works as expected (fails without libatomic, succeeds with the flag added):
```
36-150% /opt/local/bin/gcc-mp-14 -std=c++17 -o require_libatomic require_libatomic.cc
Undefined symbols:
  "___atomic_fetch_add_8", referenced from:
      __ZNSt13__atomic_baseIxEppEv in ccpntpHJ.o
  "___atomic_load_8", referenced from:
      __ZNKSt13__atomic_baseIxEcvxEv in ccpntpHJ.o
ld: symbol(s) not found
collect2: error: ld returned 1 exit status
36-150% /opt/local/bin/gcc-mp-14 -std=c++17 -o require_libatomic require_libatomic.cc -latomic
36-150%
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -61,6 +61,7 @@ set_property( CACHE gpu_backend PROPERTY STRINGS
 
 # After color.
 include( "cmake/util.cmake" )
+include( "cmake/config.cmake" )
 
 # Recognize CTest's BUILD_TESTING flag. (Quotes required.)
 if (NOT "${BUILD_TESTING}" STREQUAL "")
@@ -357,7 +358,6 @@ endif()
 
 #-------------------------------------------------------------------------------
 # Clean stale defines.h from Makefile-based build.
-message( "" )
 file( REMOVE "${CMAKE_CURRENT_SOURCE_DIR}/include/blas/defines.h" )
 
 # Include directory.
@@ -371,6 +371,8 @@ target_include_directories(
 )
 
 # OpenMP support.
+message( "" )
+message( STATUS "${bold}Looking for OpenMP${not_bold}" )
 set( openmp_lib "" )
 set( blaspp_use_openmp false )  # output in blasppConfig.cmake.in
 if (NOT use_openmp)
@@ -384,7 +386,15 @@ else()
     endif()
 endif()
 
+# If -latomic is required, add as link library to blaspp.
+message( "" )
+check_libatomic()
+if (libatomic_required)
+    target_link_libraries( blaspp PUBLIC "-latomic" )
+endif()
+
 # Get git commit id.
+message( "" )
 if (EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
     execute_process( COMMAND git rev-parse --short HEAD
                      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}

--- a/cmake/config.cmake
+++ b/cmake/config.cmake
@@ -1,0 +1,52 @@
+# Copyright (c) 2017-2024, University of Tennessee. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+# This program is free software: you can redistribute it and/or modify it under
+# the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#-------------------------------------------------------------------------------
+# Tests whether using `std::atomic` requires linking with `-latomic`
+# for 64-bit values, which is the case on some 32-bit systems.
+# Sets variable `libatomic_required`.
+#
+function( check_libatomic )
+    message( STATUS "Checking whether std::atomic requires libatomic" )
+    set( libatomic_required false )
+
+    try_compile(
+        link_result ${CMAKE_CURRENT_BINARY_DIR}
+        SOURCES
+            "${CMAKE_CURRENT_SOURCE_DIR}/config/std_atomic.cc"
+        OUTPUT_VARIABLE
+            link_output
+    )
+    debug_try_compile( "std_atomic.cc" "${link_result}" "${link_output}" )
+
+    set( label "   std::atomic links without -latomic" )
+    pad_string( "${label}" 50 label )
+    if (link_result)
+        message( "${label} ${blue} yes${plain}" )
+    else()
+        message( "${label} ${red} no${plain}" )
+
+        try_compile(
+            link_result ${CMAKE_CURRENT_BINARY_DIR}
+            SOURCES
+                "${CMAKE_CURRENT_SOURCE_DIR}/config/std_atomic.cc"
+            LINK_LIBRARIES
+                "-latomic"
+            OUTPUT_VARIABLE
+                link_output
+        )
+        debug_try_compile( "std_atomic.cc" "${link_result}" "${link_output}" )
+
+        set( label "   std::atomic requires -latomic" )
+        pad_string( "${label}" 50 label )
+        if (link_result)
+            #target_link_libraries( ${tgt} PUBLIC "-latomic" )
+            message( "${label} ${blue} yes${plain}" )
+            set( libatomic_required true )
+        else()
+            message( "${label} ${red} failed; cannot compile libatomic test${plain}" )
+        endif()
+    endif()
+endfunction()

--- a/config/config.py
+++ b/config/config.py
@@ -785,6 +785,27 @@ def gpu_blas():
 # end
 
 #-------------------------------------------------------------------------------
+def libatomic():
+    '''
+    Tests whether using `std::atomic` requires linking with `-latomic`
+    for 64-bit values, which is the case on some 32-bit systems.
+    If so, adds `-latomic` to LIBS.
+    '''
+    print_test( 'std::atomic links without -latomic' )
+    (rc, out, err) = compile_exe( 'config/std_atomic.cc' )
+    print_result( 'x', rc )
+    if (rc != 0):
+        print_test( 'std::atomic requires -latomic' )
+        env = {'LIBS': '-latomic'}
+        (rc, out, err) = compile_exe( 'config/std_atomic.cc', env )
+        print_result( 'x', rc )
+        if (rc == 0):
+            environ.merge( env )
+        else:
+            raise Error( 'cannot compile libatomic test' )
+# end
+
+#-------------------------------------------------------------------------------
 def get_package( name, directories, repo_url, tar_url, tar_filename ):
     '''
     Searches for a package, generally used for internal packages.

--- a/config/std_atomic.cc
+++ b/config/std_atomic.cc
@@ -1,0 +1,16 @@
+// Copyright (c) 2017-2024, University of Tennessee. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include <atomic>
+#include <cstdint>
+
+int main( int argc, char** argv )
+{
+    std::atomic<std::int64_t> x = 0;
+    for (int i = 1; i < argc; ++i) {
+        ++x;
+    }
+    return x;
+}

--- a/configure.py
+++ b/configure.py
@@ -57,6 +57,9 @@ def main():
    #config.prog_cxx_flag( '-Wconversion' )
    #config.prog_cxx_flag( '-Werror' )
 
+    print_header( 'Libraries' )
+    config.libatomic()
+
     config.openmp()
 
     config.lapack.blas()

